doc/js-perspective-on-c.md

```diff
@@ -0,0 +1,108 @@
+
+In modern JS we generally pass in what is needed inside the function context, return back the values that result and if an error occurs it is thrown.  That pattern generally holds true for both sync an async functions.
+
+`C` stack-allocation semantics, and the resulting variable scoping, often result in a pattern where both input and output arguments are passed and then the function either returns `void` or `int` (to denote the error code where `0` represents no error) to show if valid execution occurred.  Generally one sees variables allocated for both inputs and return(s) in the calling context and then those variables are passed into the function.  Many times the function call is placed in a conditional to check execution status.
+
+```c
+KZGSettings settings; // this is the output of load_trusted_setup_file()
+FILE *file_handle = fopen(******, "r"); // this is the input of load_trusted_setup_file()
+C_KZG_RET ret = load_trusted_setup_file(&settings, file_handle);
+
+if (fclose(file_handle) != 0) {
+    if (ret == C_KZG_OK) {
+        free_trusted_setup(&(data->settings));
+    }
+}
+
+if (ret != C_KZG_OK) {
+    return 1;
+}
+
+return 0;
+```
+
+There is one more paradigm that is worth explicitly stating.  Note in this example that `fclose` is in the conditional and `load_trusted_setup_file` is not.  `load_trusted_setup_file` is a function where the return value needs to be handled for either valid or invalid execution (ie valid but if fclose fails need to `free_trusted_setup` and for invalid execution the function returns `1`).  `fclose` however only requires that invalid execution requires further work.
+
+There are a few challenges that emerge from the way that `c` functions are structured.  In very large code-bases its very difficult to track allocations and de-allocations because they do not always occur in the same functions.
+
+In `c++` everything from above is true and there are some additional paradigms that should be presented.  Object have higher complexity for initialization/clean-up and bigger sizes.  Those "enhancements" and the `class` paradigm are where most of the complexity in `c++` (relative to `c`) comes in.  The trade of for the complexity is RAII.  Or i suppose RAII emerged because of the complexity and it is a very nice paradigm for large and complex systems.
+
+RAII helps to overcome a lot of the challenges that emerge with allocations and de-allocations that occur in different locations in the codebase.  Most of the "magic" that happens in `c++` is the compiler filling in "defaults" so that the class structure isn't overly verbose.  It is not necessary to write functions for constructors, destructors, copy semantics nor move semantics.  They get "auto" generated for us, sometime with disastrous consequences.
```

Comment:
```suggestion
RAII helps to overcome a lot of the challenges that emerge with allocations and de-allocations that occur in different locations in the codebase.  Most of the "magic" that happens in `C++` is the compiler filling in "defaults" so that the class structure isn't overly verbose.  It is not necessary to write functions for constructors, destructors, copy semantics nor move semantics.  They get "auto" generated for us, sometimes with disastrous consequences.
```

Reply:
Updated.